File: backpack/518.coin-change-2.rb
# @param {Integer} amount
# @param {Integer[]} coins
# @return {Integer}
def change(amount, coins)
  dp = Array.new(amount + 1, 0)
  dp[0] = 1

  (1..amount).to_a.each do |amt|
    coins.each do |coin|
      next if amt < coin

      dp[amt] += dp[amt - coin]
    end
  end

  dp[amount]
end

File: backpack/518.coin-change-2.py
import unittest


class Solution:
    def cmbChange(self, coins, amount):
        DP = [0 for _ in range(amount + 1)]
        DP[0] = 1

        for coin in coins:
            print(coin)
            for i in range(1, amount + 1):
                if i >= coin:
                    DP[i] += DP[i-coin]
            print(DP)
        return DP[amount]

class SolutionTest(unittest.TestCase):
    def testCmbChange(self):
        sol = Solution()
        sol.cmbChange([1, 2, 5], 5)

if __name__ == "__main__":
    unittest.main()
